### ISSUE TITLE
Async Step Definitions

### ISSUE THREAD

**Issue by jeduden** [pytest-dev/pytest-bdd#223]
is there away to use async step definitions with pytest-bdd ?

For a step definition like:
`
@when('i send cucumbers ')
async def i_send_cucumbers(loop):
      pass
`
I see the warning:
`
pytest_bdd/scenario.py:137: RuntimeWarning: coroutine 'i_send_cucumbers' was never awaited
`

**Comment by olegpidsadnyi** [comment]
what about async then? does it even make sense?

**Comment by olegpidsadnyi** [comment]
will this work for you?


async def send_cucumbers():
    pass


@when('i send cucumbers ')
def i_send_cucumbers(loop):
    loop.run_until_complete(send_cucumbers)

**Comment by jeduden** [comment]
This works. However, in my case all step definitions are async (and can call multiple times into async functions). so this would mean a lot of calls to run_until_complete.
As a work-around I have defined a decorator that schedules the function on the event loop.

A generic solution would be to have the ability to easily wrap any step definition with a function. (similar to the before and after hooks).

And then it would be great if support of async functions would be provided out of the box. 

Especially, after reading the discussion about implicit and explicit eventloops in the thread here: https://groups.google.com/forum/#!msg/python-tulip/yF9C-rFpiKk/tk5oA3GLHAAJ seems to be tending towards implicit loops for python code that runs on 3.5.3+ and/or 3.6+. 

So code that explicitly calls run_until_complete to run async functions will look more and more awkward.

**Comment by olegpidsadnyi** [comment]
@jeduden this means pytest-bdd has to depend on yet another library asyncio? I don't really see how a gherkin scenario can by async. If the whole point that it has not to break i think you need to extend your testing suite with some kind of support for async functions (decorator is a good idea).

I think semantically you can't really represent async process in step-by-step imperative Gherkin. Therefore it should be explicit. If you have to initiate few async messages - they should stand behind one When step that describes this exercise in a form that humans understand. Gherkin is not a programming language, it is the way to describe steps to humans which can't do async computation in their brains.

What do you think, @bubenkoff ?

**Comment by bubenkoff** [comment]
well, the async question comes down to the pytest, not specifically to pytest-bdd.
And for pytest's dependency injection, it's not realistic that it starts to support async fixture definitions anytime soon. Also for tests, while it sounds cool there seems to be a little win to have async fixtures, simply because fixtures should be fast enough, and then it should not matter much if you optimise dependency graph in a way that you run fixtures in parallel for some parts of the graph.
It worth effort though to add the 'async clause' to the documentation mentining the workaround to cut the async point on the fixture definition by waiting for async function to finish

**Comment by olegpidsadnyi** [comment]
@jeduden could you show your decorator implementation for when step that is awaiting?

**Comment by bubenkoff** [comment]
FYI: there's a helper apparently to minimize the efforts: https://pypi.python.org/pypi/pytest-asyncio

**Comment by jeduden** [comment]
Support on pytest side is fine. We are using aiohttp test-support and with pytest-asyncio 0.8.0 we can also the helpers from that package.

However, these helpers work on scenario level. 

The step_definitions are one level below, and since the calling code here: https://github.com/pytest-dev/pytest-bdd/blob/master/pytest_bdd/scenario.py#L137 
is not checking if step_func is async. i.e. it only supports synchronous step functions.

In order to fix this, we currently need to wrap all steps with decorators like this:
```
 def sync(func):
      @wraps(func)
      def synced_func(*args, **kwargs):
          loop = kwargs.get("loop")
          if not loop:
              raise Exception("Need loop fixture to make function sync")
          return loop.run_until_complete(func(*args, **kwargs))
      return synced_func
```

**Comment by jeduden** [comment]
@olegpidsadnyi declaring a function async doesn't mean it is not imperative. Async functions are imperative like synchronous functions. However, with await / yield from you explicitly define points where the execution of other scheduled coroutines is allowed. 

In order to schedule the parallel execution of multiple asynchronous functions you would use helpers like asyncio.gather ( see https://docs.python.org/3/library/asyncio-task.html#example-parallel-execution-of-tasks )

The reason why we need to declare our step_definitions async is because we are using the aiohttp client inside to perform networking calls as part of our step definitions. 
Furthermore, the rest of the code base is completely async, hence we think it is only natural that also step_defintions are written with async.

**Comment by bubenkoff** [comment]
As i see from the pytest-asyncio code, it just awaits for every 'async' fixture, so there's no real parallelism possible between the fixtures
And how do you mean pytest upports async fixtures in a 'proper' way? Can't really see that

**Comment by bubenkoff** [comment]
But i do see your point about the need of the wrapper everywhere - it sucks.
Looks like the only way to avoid that is to depend on pytest-asyncio and use it's helpers directly in the pytest-bdd

**Comment by jeduden** [comment]
@bubenkoff even that is not possible since there is no hook that allows me to insert a custom decorator on the functions.

even if pytest-bdd doesn't support async step definitions out of the box. A hook for wrapping a step definitions would help to reduce duplication.

**Comment by bubenkoff** [comment]
but I meant to change pytest-bdd itself to support that automatically

**Comment by bubenkoff** [comment]
are you up for making a PR which will add `pytest-asyncio` as a dependency and automatically use it to resolve async step definitions, if they are async?

**Comment by jeduden** [comment]
@bubenkoff In general, I am happy to make a PR. One question regarding the pytest-asyncio dependency. You mention it because we need to have a library to provide the loop fixture, right ?

**Comment by bubenkoff** [comment]
@jeduden yes, also to keep as much `core` async stuff as possible in a single plugin (pytest-asyncio)

**Comment by jeduden** [comment]
@bubenkoff i did a quick check on the current test-suite; how i can best write some this for this new type of possible step functions. do you have a pointer for me where i best can add tests ? Or should i create a complete new file ?

**Comment by bubenkoff** [comment]
You can put a new file here tests/steps/test_async.py copying the approach of tests/steps/test_unicode.py
and replacing definitions with async ones:
```
@given
async def ...

@when
async def ...

@then
async def ...
```

**Comment by s0undt3ch** [comment]
https://github.com/pytest-dev/pytest-bdd/pull/221

**Comment by vodik** [comment]
For what its worth, I took the approach with #221 and then added that hook implementation as a separate pytest plugin that I can install as necessary.

**Comment by Victor-Savu** [comment]
@vodik 
> added that hook implementation as a separate pytest plugin

Is your pytest plugin available anywhere? Do you plan on maintaining it? My gut feeling is that integrating the functionality into `pytest-bdd` would be the path of least maintenance burden, but that is up to the maintainers I suppose.

**Comment by Victor-Savu** [comment]
First off, thank you so much for this awesome library! I was hoping I can contribute my experience with `asyncio`.

@olegpidsadnyi sad [here](https://github.com/pytest-dev/pytest-bdd/issues/223#issuecomment-331870328):
> this means pytest-bdd has to depend on yet another library asyncio?
Depends on how you look at it. `asyncio` is part of the standard library, same as `json` or `datetime`. So users wouldn't need to install anything special.

> I think semantically you can't really represent async process in step-by-step imperative Gherkin.
Agreed, *but* the cool thing is that despite the name, `async/await` doesn't necessarily mean that the operations happen in arbitrary order. `await` just means that `async` operations suspend the current execution thread (not system thread, mind you) and execution is resumed once they are completed, in the same order as specified in the function. This way, the semantics of the operations are unchanged happen in a step-by-step imperative mode.
Please let me know if you would like to have a conversation about this topic. I'm always happy to talk about `async/await` :)

**Comment by rgreinho** [comment]
Behave 1.2.6 add some decorators to [Testing asyncio Frameworks ](https://behave.readthedocs.io/en/latest/new_and_noteworthy_v1.2.6.html#testing-asyncio-frameworks). 

That may be inspirational to implement something similar for pytest-bdd.

**Comment by pjz** [comment]
Is there a guide somewhere to using pytest-bdd with pytest-asyncio ?  I'm not sure how to make them play nicely together and actually execute my bdd tests.

**Comment by DjaPy** [comment]
@bubenkoff Hi! Are you planning to close this [PR](https://github.com/pytest-dev/pytest-bdd/pull/349)?

**Comment by mathew-jithin** [comment]
Just shooting this question as I am stuck with the same issue for handling async step definitions as pytest-asyncio cannot be used here.

Is there some update or any way to handle the same as of now? Hope you will please help if some new changes/ways to handle are present.

@bubenkoff

**Comment by jruizaranguren** [comment]
Most of web development nowadays is moving to `async`  frameworks (for instance [fastapi](https://fastapi.tiangolo.com/)). Testing this frameworks typically involves running async app and also use an async testclient.
Support for async pytest-bdd step definitions would definitely help in further adoption in these kind of projects.

**Comment by bubenkoff** [comment]
@DjaPy not for me to decide, I've requested a review from @youtux

**Comment by DjaPy** [comment]
I forked the project and applied [PR](https://github.com/pytest-dev/pytest-bdd/pull/349). I was able to take advantage of asynchronous tests. But as time passes, I see that this is unnecessary. If you write integration tests, then nothing prevents you from using [requests](https://requests.readthedocs.io/en/master/) as a client. I will support the voiced [idea](https://github.com/pytest-dev/pytest-bdd/issues/223#issuecomment-331860204) that adding asynchrony is redundant and solves other problem.

**Comment by detoyz** [comment]
@DjaPy but what about if your integration tests are part of your whole asyncio project with lots of async tests? Isn't it's better to just run `pytest` on the whole project and view all test results at once? I thought that's one of the key advantages of pytest-bdd. Otherwise, if you write your bdd tests fully isolated from other project test structure (async fixtures for example), what's the benefit then of using pytest-bdd instead of let's say behave?
Currently, I actually do as you suggested, by separating bdd tests (using requests in them) and other tests. However again, for local development, for CI/CD I always need to run two commands instead of one. I mean it's not probably a "must-have" feature, but definitely "nice-to-have" for pytest-bdd.

**Comment by DjaPy** [comment]
> I thought that's one of the key advantages of pytest-bdd. Otherwise, if you write your bdd tests fully isolated from other project test structure (async fixtures for example), what's the benefit then of using pytest-bdd instead of let's say behave?

This is an argument. When you're used to writing in pytest and you don't want to dive into behave. I'll choose pytest-bdd. Yet here another problem is solved. The problem of infrastructure.

**Comment by Limonkufu** [comment]
Do we have a workaround getting async to work with pytest-bdd with aiohttp with `async with`? I was wrapping my async calls with `loop.run_until_complete' as suggested above but it just means that I have to define every fixture twice and then wrap them just to get it to work.

**Comment by youtux** [comment]
I made a POC at https://github.com/pytest-dev/pytest-bdd/pull/629. It only implements the execution of async step functions.
If you need to use async fixture, you can do that via a plugin like `pytest-asyncio`.

The implementation of the fix seems so trivial that I'm not sure I want to include it in pytest-factoryboy. Anybody can make a decorator that converts async functions to sync:

```python
def async_to_sync(fn):
    """Convert async function to sync function."""

    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        return asyncio.run(fn(*args, **kwargs))

    return wrapper
```

and apply it to their async steps:

```python
@given("there is a cucumber", target_fixture="cucumber")
async def _():
    return 42
```

**Comment by Cito** [comment]
Thanks @youtux. This workaround was actually already suggested [6 years ago](https://github.com/pytest-dev/pytest-bdd/issues/223#issuecomment-332084037).

Unfortunately it is not compatible with using pytest-asyncio for running async fixtures. When the test function uses an async fixture via pytest-asyncio, then an event loop is already running, but the `asyncio.run` call will create a *different* event loop and not run the test code inside the already running event loop used with the async fixture. This may cause the test to hang.

**Comment by Cito** [comment]
I am now using the following decorator for async steps, which works well with pytest-asyncio:

```python
def async_step(step):
    """Convert an async step function to a normal one."""

    signature = inspect.signature(step)
    parameters = list(signature.parameters.values())
    has_event_loop = any(parameter.name == "event_loop" for parameter in parameters)
    if not has_event_loop:
        parameters.append(
            inspect.Parameter("event_loop", inspect.Parameter.POSITIONAL_OR_KEYWORD)
        )
        step.__signature__ = signature.replace(parameters=parameters)

    @wraps(step)
    def run_step(*args, **kwargs):
        loop = kwargs["event_loop"] if has_event_loop else kwargs.pop("event_loop")
        return loop.run_until_complete(step(*args, **kwargs))

    return run_step
````

This can be applied in the same way:
```python
@given("there is a cucumber", target_fixture="cucumber")
@async_to_sync
async def there_is_a_cucumber():
    return 42
```

It works like this: The `async_step` decorator first checks whether the step function already requests the `event_loop` fixture of pytest-asyncio. If not, it requests it by adding it to the signature of the step function. The innner `run_step` function can now be sure that the `event_loop` is passed as a parameter, fetches it from there and runs the async function in that event loop.

**Comment by youtux** [comment]
Do you actually need all that signature machinery? Wouldn't this work as well?

```python
import asyncio, functools

def async_to_sync(fn):
    """Convert async function to sync function."""

    @functools.wraps(fn)
    def wrapper(event_loop, *args, **kwargs):
        return event_loop.run_until_complete(fn(*args, **kwargs))

    return wrapper
```

**Comment by Cito** [comment]
> Do you actually need all that signature machinery? Wouldn't this work as well?

See my other [comment ](https://github.com/pytest-dev/pytest-bdd/issues/223#issuecomment-1647858564).

**Comment by youtux** [comment]
Note that my last example is different from what I originally suggested.

**Comment by Cito** [comment]
> Note that my last example is different from what I originally suggested.

Sorry, yes, you're right. That simplified code should work, but does not cover the case that the original step function takes the event loop fixture as a parameter.

**Comment by youtux** [comment]
Ah yes you're right

**Comment by seifertm** [comment]
I come back to this issue from time to time, because there's a corresponding [issue in pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio/issues/195).

My current understanding is that pytest-asyncio makes async step definitions hard to implement, because it assumes by default that each async test item runs in their own event loop. In pytest-bdd, however, you want to run multiple steps inside the same loop.

Up until pytest-asyncio v0.21, the only way to have an asyncio event loop scope other than function-scope is to reimplement the _event_loop_ fixture. This leads to a bunch of problems and the plan is to deprecated and remove this functionality.

As a replacement for event loop fixture overrides, pytest-asyncio will provide the _asyncio_event_loop_ mark (see https://github.com/pytest-dev/pytest-asyncio/pull/620). When the mark is added to a test class or to a module, pytest-asyncio will provide an asyncio event loop with the respective scope and run all tests under the mark in that scoped loop. It should also run any async fixture inside that same loop (although this needs more testing). Here's an example from the docs:
```python
import asyncio

import pytest
import pytest_asyncio

@pytest.mark.asyncio_event_loop
class TestClassScopedLoop:
    loop: asyncio.AbstractEventLoop

    @pytest_asyncio.fixture
    async def my_fixture(self):
        TestClassScopedLoop.loop = asyncio.get_running_loop()

    @pytest.mark.asyncio
    async def test_runs_is_same_loop_as_fixture(self, my_fixture):
        assert asyncio.get_running_loop() is TestClassScopedLoop.loop
```

I'd be happy to provide a pre-release version to play around with this feature. Alternatively, you can just pip install  from 
git, e.g. `pip install git+https://github.com/pytest-dev/pytest-asyncio@106fa545a659a7e6a936b0f53d9d184287be8a13`

Do you think this would help solve this issue?

**Comment by seifertm** [comment]
I imagine that pytest-bdd applies the mark under the hood via `pytest.Item.add_marker`. The pytest-bdd user shouldn't have to get in touch with the marker in my opinion.

That being said, the pytest-asyncio-0.22.0 release has been yanked. It's currently unclear if markers will be the way to go.

**Comment by ShedPlant** [comment]
As a much simpler workaround, you can (at least with `pytest-bdd` `7.3.0`, `pytest-asyncio` `0.23.5`) define in two steps:

```python
@fixture
async def provisioned_thing(client, thing):
    await client.create(thing)
    return thing

@given("a thing")
def thing(provisioned_thing):
    return provisioned_thing
```

**Comment by vladyslav2** [comment]
Its already been 5 years since this ticket started

I just wonder, does no one using async with bdd? or just pytest-bdd aren't very popular amount async frameworks?

Cause all current solutions seems like a workarounds ...

**Comment by jruizaranguren** [comment]
@vladyslav2 I think many people use `async` with pytest-bdd, but probably not for `step` definitions.  We use async fixtures that are injected in step definitions so we can re-use production code. But we have not found much benefit in having the `steps` themselves being `async`. In order to speed up test execution, and also have better random execution properties, we use `pytest-xdist`. 

Which benefits do you expect from having async step definitions?

**Comment by Cito** [comment]
>Which benefits do you expect from having async step definitions?

I cannot speak for @vladyslav2, but e.g. you may only have an async driver to check objects in a database or object storage when you do whitebox testing.

**Comment by vladyslav2** [comment]
@jruizaranguren I believe your approach works fine for simple cases where you need to do a request and check response. 
but once you get more into the woods, you want to reuse part of what you already have but its all mostly async.

Real case scenario:
- User do an action XXX
- You want to check a response
- You want to get data from database to make sure your business logic is correct, and its async
- You want to check if email/web notification fired, and its async
- You want to check if some 3rd party api was executed, and your http fuctions also async
- etc

**Comment by jruizaranguren** [comment]
Sure, @vladyslav2, I had plenty of occasions where I wanted to do whitebox testing (as @Cito mentioned), and I have spent a lot of time trying to make async play well with BDD tests.  But in the end I found that fighting the tool was wasting my time without a meaningful benefit. I do have really large tests suites:

 My current take (good enough, might not be the best):

* use BDD for blackbox testing: IMHO specifications implementation should know nothing about implementation details
* ensure performance and randomness: use pytest.xdist
* use  `pytest_asyncio.fixture` for setup injected in given clauses if required
* use regular tests for whitebox testing

> You want to check if email/web notification fired, and its async
> You want to check if some 3rd party api was executed, and your http fuctions also async

Those sound like integration testss that may be implemented with HTTP request-reply. I understand your approach might vary, specially if your are implementing a service with almost not computing, only I/O.

**Comment by zbyte64** [comment]
What worked for me (with pytest-asyncio installed):

```
def async_to_sync(fn):
    """Convert async function to sync function."""

    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        c = fn(*args, **kwargs)
        loop = asyncio.get_event_loop()
        return loop.run_until_complete(c)

    return wrapper

@given(parsers.parse('a device "{device_id}" is registered to restaurant "{restaurant_id}"'))
@async_to_sync
async def device_registered(db_session, device_id, restaurant_id):
     ....
```
